File: assignment_03/src/main.rs
use std::str::FromStr;

#[derive(Debug, Clone)]
struct Change {
    total: usize,
    amounts: Vec<usize>,
}

impl Change {
    fn new(dnmn: usize) -> Self {
        let amounts = vec![0; dnmn];

        Change {
            total: 0,
            amounts,
        }
    }
}

#[derive(Debug, Clone)]
enum State {
    Set(Change),
    Unset,
    Invalid,
}

impl State {
    fn ref_change(&self) -> Option<&Change> {
        match self {
            State::Set(v) => Some(&v),
            _ => None
        }
    }
}

impl Default for State {
    fn default() -> Self {
        State::Unset
    }
}

fn parse_line_fill<T>(line: &str, list: &mut Vec<T>) -> bool
where
    T: FromStr
{
    let split = line.split(' ');

    for value in split {
        if value.is_empty() {
            continue;
        }

        let Ok(parsed) = value.parse() else {
            return false;
        };

        list.push(parsed);
    }

    true
}

fn parse_line<T>(line: &str) -> Option<Vec<T>>
where
    T: FromStr
{
    let mut rtn = Vec::new();

    if parse_line_fill(line, &mut rtn) {
        Some(rtn)
    } else {
        None
    }
}

fn main() {
    let mut method_to_use = 0;
    let mut run_checks = false;

    for arg in std::env::args() {
        match arg.as_str() {
            "--run-checks" => {
                run_checks = true;
            },
            "--bottom-up-2" => {
                method_to_use = 1;
            },
            "--top-down" => {
                method_to_use = 2;
            },
            _ => {}
        }
    }

    let mut total_denominations: usize = 0;
    let mut total_checks: usize = 0;
    let mut max_size: usize = 0;
    let mut lines = std::io::stdin().lines();
    let mut denominations: Vec<usize> = Vec::new();
    let mut checks: Vec<usize> = Vec::new();

    {
        let Some(check) = lines.next() else {
            panic!("no change line specified");
        };

        let change_line = check.expect("failed to read input from stdin");

        let Some(change_data) = parse_line::<usize>(&change_line) else {
            panic!("invalid change line provided: \"{}\"", change_line);
        };

        if change_data.len() < 2 {
            panic!("too few change values: \"{}\"", change_line);
        }

        if change_data[0] == 0 {
            panic!("amount of denominations specified is invalid: {}", change_data[0]);
        }

        total_denominations = change_data[0];
        total_checks = change_data[0];
    }

    {
        let Some(check) = lines.next() else {
            panic!("no denominations specified");
        };

        let check_line = check.expect("failed to read input from stdin");

        if !parse_line_fill(&check_line, &mut denominations) {
            panic!("invalid denominations line provided: \"{}\"", check_line);
        }
    }

    if denominations.len() != total_denominations {
        panic!("denominations provided nodes not match the specified amount");
    }

    while let Some(line) = lines.next() {
        let line_check = line.expect("failed to read input from stdin");

        let Ok(value): Result<usize, _> = line_check.parse() else {
            panic!("invalid check value provided: \"{}\"", line_check);
        };

        if value > max_size {
            max_size = value;
        }

        checks.push(value);
    }

    if checks.len() != total_checks {
        panic!("checks provided does not match the specified amount\n");
    }

    if run_checks {
        for dnmn in &denominations {
            print!("{} ", dnmn);
        }

        println!("");
    }

    match method_to_use {
        0 => {
            let mut running_count = vec![0; denominations.len()];
            let mut memorized_bottom_up = vec![(None, 0); max_size + 1];

            calc_bottom_up(&denominations, &mut memorized_bottom_up);

            for value in &checks {
                let mut index = *value;

                loop {
                    let (Some(v), lu) = &memorized_bottom_up[index] else {
                        break;
                    };

                    if *v == 0 {
                        break;
                    }

                    running_count[*lu] += 1;
                    index = index - denominations[*lu];
                }

                for count in &running_count {
                    print!("{} ", count);
                }

                println!("");

                running_count.fill(0);
            }
        },
        1 => {
            let mut memorized_bottom_up = vec![State::Unset; max_size + 1];

            calc_bottom_up_alt(&denominations, &mut memorized_bottom_up);

            for value in &checks {
                match &memorized_bottom_up[*value] {
                    State::Set(change) => {
                        if run_checks {
                            print!("{} => {}: ", value, change.total);

                            let mut count = 0;

                            for dnmn_index in 0..denominations.len() {
                                print!("{} ", change.amounts[dnmn_index]);

                                count += denominations[dnmn_index] * change.amounts[dnmn_index];
                            }

                            print!("= {}", count);

                            if count != *value {
                                println!(" invalid");
                            } else {
                                println!(" valid");
                            }
                        } else {
                            for dnmn in &change.amounts {
                                print!("{} ", dnmn);
                            }

                            println!("");
                        }
                    },
                    State::Invalid => {
                        for _ in &denominations {
                            print!("0 ");
                        }

                        println!("");
                    },
                    State::Unset => unreachable!(),
                }
            }
        },
        2 => {
            let mut memorized_top_down = vec![State::Unset; max_size + 1];

            denominations.reverse();

            for value in &checks {
                std::thread::scope(|scope| {
                    let result = std::thread::Builder::new()
                        .name(format!("top_down_{}", value))
                        .stack_size(512 * 1024 * 1024)
                        .spawn_scoped(scope, || {
                            calc_top_down(*value, &denominations, &mut memorized_top_down, 1);
                        });

                    if let Err(err) = result {
                        println!("thread error: {:#?}", err);
                    }
                });

                match &memorized_top_down[*value] {
                    State::Set(change) => {
                        if run_checks {
                            print!("{} => {}: ", value, change.total);
                            let mut count = 0;

                            for dnmn_index in (denominations.len() - 1)..=0 {
                                print!("{} ", change.amounts[dnmn_index]);

                                count += denominations[dnmn_index] * change.amounts[dnmn_index];
                            }

                            println!("= {}", count);
                        } else {
                            for dnmn in change.amounts.iter().rev() {
                                print!("{} ", dnmn);
                            }

                            println!("");
                        }

                        for dnmn in change.amounts.iter().rev() {
                            print!("{} ", dnmn);
                        }

                        println!("");
                    },
                    State::Invalid => {
                        for _ in &denominations {
                            print!("0 ");
                        }

                        println!("");
                    },
                    State::Unset => unreachable!(),
                }
            }
        },
        _ => unreachable!(),
    }
}

fn calc_top_down(change: usize, denominations: &[usize], memorized: &mut [State], context: usize) -> State {
    if change == 0 {
        #[cfg(debug_assertions)]
        println!("{} {} base case", context, change);

        return State::Set(Change::new(denominations.len()));
    }

    if let Some(found) = memorized[change].ref_change() {
        #[cfg(debug_assertions)]
        println!("{} {} memorized: {:?}", context, change, found);

        return State::Set(found.clone());
    }

    let mut lowest = State::Invalid;

    for dnmn in 0..denominations.len() {
        if denominations[dnmn] > change {
            #[cfg(debug_assertions)]
            println!("{} {} < dnmn {}", context, change, denominations[dnmn]);

            continue;
        }

        #[cfg(debug_assertions)]
        println!("{} {} sub: {}", context, change, denominations[dnmn]);

        let mut result = match calc_top_down(change - denominations[dnmn], denominations, memorized, context + 1) {
            State::Set(change) => change,
            State::Unset => unreachable!(),
            State::Invalid => continue,
        };

        lowest = match lowest {
            State::Set(curr) => {
                if curr.total > result.total {
                    result.total += 1;
                    result.amounts[dnmn] += 1;

                    #[cfg(debug_assertions)]
                    println!("{} {} updating lowest: {:?} prev: {:?}", context, change, result, curr);

                    State::Set(result)
                } else {
                    State::Set(curr)
                }
            },
            State::Unset => unreachable!(),
            State::Invalid => {
                result.total += 1;
                result.amounts[dnmn] += 1;

                #[cfg(debug_assertions)]
                println!("{} {} setting lowest: {:?}", context, change, result);

                State::Set(result)
            }
        };
    }

    memorized[change] = lowest;

    #[cfg(debug_assertions)]
    println!("{} {} result: {:?}", context, change, memorized[change]);

    memorized[change].clone()
}

fn calc_bottom_up_alt(denominations: &[usize], memorized: &mut [State]) {
    let dnmn_len = denominations.len();

    memorized[0] = State::Set(Change::new(dnmn_len));

    for index in 1..memorized.len() {
        let mut min = State::Invalid;

        #[cfg(debug_assertions)]
        println!("{}", index);

        for dnmn in 0..dnmn_len {
            #[cfg(debug_assertions)]
            print!("├─── {}", denominations[dnmn]);

            if denominations[dnmn] > index {
                #[cfg(debug_assertions)]
                println!(" skip");

                break;
            }

            let Some(cmp) = memorized[index - denominations[dnmn]].ref_change() else {
                #[cfg(debug_assertions)]
                println!(" no change for previous");

                continue;
            };

            #[cfg(debug_assertions)]
            print!(" {:?}", cmp);

            min = match min {
                State::Set(curr) => {
                    let mut rtn = cmp.clone();
                    rtn.total += 1;
                    rtn.amounts[dnmn] += 1;

                    #[cfg(debug_assertions)]
                    print!(" {} < {}", curr.total, rtn.total);

                    if curr.total <= rtn.total {
                        #[cfg(debug_assertions)]
                        println!(" no update");

                        State::Set(curr)
                    } else {
                        #[cfg(debug_assertions)]
                        println!(" updating to {:?}", rtn);

                        State::Set(rtn)
                    }
                },
                State::Unset => unreachable!(),
                State::Invalid => {
                    let mut rtn = cmp.clone();
                    rtn.total += 1;
                    rtn.amounts[dnmn] += 1;

                    #[cfg(debug_assertions)]
                    println!(" setting to {:?}", rtn);

                    State::Set(rtn)
                }
            };
        }

        #[cfg(debug_assertions)]
        println!("└─── memorizing {:?}", min);

        memorized[index] = min;
    }
}

fn calc_bottom_up(dnmn: &[usize], mem: &mut [(Option<usize>, usize)]) {
    mem[0] = (Some(0), 0);

    let mut min = None;
    let mut last_used = 0;

    for i in 1..mem.len() {
        min = None;
        last_used = 0;

        for d in 0..dnmn.len() {
            if dnmn[d] > i {
                continue;
            }

            let (Some(check), _) = &mem[i - dnmn[d]] else {
                continue;
            };

            let cmp = check + 1;

            min = Some(if let Some(min) = min {
                if min <= cmp {
                    min
                } else {
                    last_used = d;
                    cmp
                }
            } else {
                last_used = d;
                cmp
            });
        }

        mem[i] = (min, last_used);
    }
}

File: assignment_03/main.cpp
#include <iostream>
#include <sstream>
#include <string>
#include <tuple>
#include <stdexcept>
#include <vector>
#include <utility>
#include <map>
#include <set>
#include <queue>

struct Change {
    bool is_set = false;
    std::size_t count = 0;
    std::size_t last_used = 0;
    std::size_t possible = 0;
};

/**
 * parses a given string to extract a list of integers
 */
void get_int_list(std::string &line, std::vector<int> &list) {
    std::size_t processed;
    std::size_t total_processed;
    int parsed;
    const char *str = line.c_str();

    while (true) {
        try {
            parsed = std::stoi(str, &processed);
        } catch (std::invalid_argument const& err) {
            return;
        } catch (std::out_of_range const& err) {
            return;
        }

        total_processed += processed;

        if (total_processed == line.length()) {
            return;
        }

        str += processed;

        list.push_back(parsed);
    }
}

/**
 * converts an int to a size_t
 */
bool int_to_size_t(int v, std::size_t &to_set) {
    if (v < 0) {
        return false;
    }

    to_set = (std::size_t)((unsigned)v);

    return true;
}

/**
 * calculates all values in mem with the given denominations
 */
void calc_bottom_up(std::vector<std::size_t> dnmn, std::vector<Change> &mem) {
    mem[0].is_set = true;
    mem[0].count = 0;
    mem[0].last_used = 0;

    bool is_set;
    std::size_t count;
    std::size_t last_used;
    std::size_t cmp;
    std::size_t dnmn_index;

    for (std::size_t index = 1; index < mem.size(); ++index) {
        is_set = false;
        count = 0;
        last_used = 0;

        for (dnmn_index = 0; dnmn_index < dnmn.size(); ++dnmn_index) {
            if (dnmn[dnmn_index] > index) {
                continue;
            }

            if (!mem[index - dnmn[dnmn_index]].is_set) {
                continue;
            }

            cmp = mem[index - dnmn[dnmn_index]].count + 1;

            if (is_set) {
                if (count > cmp) {
                    last_used = dnmn_index;
                    count = cmp;
                }
            } else {
                last_used = dnmn_index;
                count = cmp;
                is_set = true;
            }
        }

        mem[index].is_set = is_set;
        mem[index].count = count;
        mem[index].last_used = last_used;
    }
}

int main(int argc, char *argv[]) {
    std::string line;
    std::vector<int> line_result;
    std::vector<std::size_t> denominations;
    std::vector<std::size_t> checks;
    std::size_t denomination_count;
    std::size_t check_count;
    std::size_t conv;
    std::size_t max_size = 0;

    if (std::getline(std::cin, line)) {
        get_int_list(line, line_result);

        if (line_result.size() != 2) {
            std::cout << "invalid change line provided: \"" << line << "\"\n";
            return 1;
        }

        if (!int_to_size_t(line_result[0], denomination_count)) {
            std::cout << "amount of denominations specified is invalid: " << line_result[0] << "\n";
            return 1;
        }

        if (!int_to_size_t(line_result[1], check_count)) {
            std::cout << "amount of checks specified is invalid: " << line_result[1] << "\n";
            return 1;
        }
    }

    if (std::getline(std::cin, line)) {
        line_result.clear();
        get_int_list(line, line_result);

        if (line_result.size() == 0) {
            std::cout << "no denominations specified\n";
            return 1;
        }

        for (int v : line_result) {
            if (!int_to_size_t(v, conv)) {
                std::cout << "invalid denominations line provided: \"" << line << "\"\n";
                return 1;
            }

            denominations.push_back(conv);
        }

        if (denominations.size() != denomination_count) {
            std::cout << "denominations provided does not match the specified amount\n";
            return 1;
        }
    }

    for (; std::getline(std::cin, line);) {
        line_result.clear();

        get_int_list(line, line_result);

        if (line_result.size() == 0) {
            std::cout << "no check amount was specified\n";
            return 1;
        }

        if (!int_to_size_t(line_result[0], conv)) {
            std::cout << "invalid check value provided: \"" << line << "\"\n";
            return 1;
        }

        if (conv > max_size) {
            max_size = conv;
        }

        checks.push_back(conv);
    }

    if (checks.size() != check_count) {
        std::cout << "checks provided does not match the specified amount\n";
        return 1;
    }

    max_size += 1;

    std::size_t index;
    std::size_t running_index;
    Change empty;
    std::vector<std::size_t> running(denominations.size());
    std::vector<Change> memorized(max_size, empty);

    calc_bottom_up(denominations, memorized);

    for (std::size_t value : checks) {
        index = value;

        while (true) {
            if (!memorized[index].is_set) {
                break;
            }

            if (memorized[index].count == 0) {
                break;
            }

            running[memorized[index].last_used] += 1;
            index = index - denominations[memorized[index].last_used];
        }

        for (running_index = 0; running_index < running.size(); ++running_index) {
            std::cout << running[running_index] << " ";
            running[running_index] = 0;
        }

        std::cout << "\n";
    }

    return 0;
}
